### DISCUSSION TITLE
How to mock custom component in unit test with new version of add_pipe using fixtures?

### DISCUSSION THREAD

**Discussion by graue70** [explosion/spaCy#7165] [Help: Coding & Implementations] [answered]
## Working setup with spaCy v2

I have a custom component (`entity_linker`) that uses an entity index (similar to a spacy kb). Since it's very large and I don't want to load it into memory every time I load the pipeline, I use a local database for the index instead. The constructor of the index class receives the path to the database. The constructor of the component receives an instance of the index class.

In my unit tests, I created a fixture for the index that depends on the [`tmp_path_factory fixture`](https://docs.pytest.org/en/stable/tmpdir.html#the-tmp-path-factory-fixture). It creates a new database in a temporary directory with test data and builds the index with that. I created another fixture which depends on the index fixture for the entity linker using the test index.
When testing my pipeline, I use the `entity_linker` fixture with `nlp.add_pipe(entity_linker)`. This works.

## Problems with spaCy v3

I am trying to upgrade to spaCy v3 and I am wondering how to do this in my case:

- Since `add_pipe` doesn't take an instance any more, I guess I need to add a factory function to the registry. Therefore I changed the `entity_linker` fixture to a language factory that returns `EntityLinker(entity_index=entity_index)`. Since the factory function is no longer a fixture that can use other fixtures, its argument `entity_index` is not known to the registry and using it yields a `ConfigValidationError` ('field required').
- I could also change the `entity_index` from fixture to misc registry function and let the other one use this (as explained in the [docs](https://spacy.io/usage/processing-pipelines#component-data-function)). However, since it uses the `tmp_path_factory`, I would surely have exactly the same problem: Using a fixture in a registry function.

So I guess my question is: **How can I use a pytest fixture (`tmp_path_factory`) in a registered function? Or can I change my architecture somehow to work around the issue?**

Here is some incomplete and simplified sample code:
```python
# entity_linker.py
class EntityIndex:
    """Stores something like a spacy kb in a local database because it's too large to be in memory"""
    def __init__(self, db_path: Path):
        self.db = self.initialize_rocksdb_database_at_db_path(db_path)

    def build_from_csv(self, csv_path: Path):
        self.go_through_csv_and_add_entities_and_aliases_and_popularity_scores_to_rocksdb_database()

    def get_ids_for_alias(self, text: str) -> List[str]:
        return self.ids_from_db_results(self.db.get(text))

    def get_entity_for_id(self, kb_id: str) -> Entity:
        return self.entity_from_db_result(self.db.get(kb_id))


class EntityLinker:
    """Custom component in a pipeline"""
    def __init__(self, entity_index: EntityIndex):
        self.entity_index = entity_index
        Doc.set_extension("entities", default=[])

    def __call__(self, doc: Doc) -> Doc:
        entities = []
        for start in range(len(doc)):
            for end in range(start + 1, len(doc) + 1):
                tokens = doc[start:end]
                kb_ids = self.entity_index.get_ids_for_alias(tokens.text)
                for kb_id in kb_ids:
                    entities.append(self.entity_index.get_entity_for_id(kb_id)
        doc._.entities = entities
        return doc
```

```python
# test_entity_linker.py
@pytest.fixture(scope="module")
def entity_index(tmp_path_factory):
    tmp_path = tmp_path_factory.mktemp(basename="entity_index", numbered=True)
    db_path = tmp_path / "db.rocksdb"
    index = EntityIndex(db_path=db_path)
    index.build_from_csv(csv_path=Path("tests/entities.csv"))
    return index


@pytest.fixture(scope="module")
def entity_linker(entity_index):
    return EntityLinker(entity_index=entity_index)


@pytest.fixture(scope="module")
def linker_pipeline(entity_linker):
    nlp = spacy.load("en_core_web_sm")
    nlp.add_pipe(entity_linker)  # This needs to be changed for spaCy v3.
    return nlp


def test_linker_sets_entities(linker_pipeline):
    doc = linker_pipeline("the world is turning")
    assert "Q2" in [ent.kb_id for ent in doc._.entities]
```

**Comment by svlandeg** [comment]
Hi! Could you share some sample code? That'll make it easier to follow ;-)

**Reply by graue70** [reply]
I added some sample code. I hope it makes the issue clear. It's a fine line between minimizing the amount of code and simplifying too much. :)

**Comment by svlandeg** [comment]
I haven't looked into detail yet with the fixtures & factories for v3, but perhaps an easy workaround for now would be to use our internal test utility `make_tempdir` in `spacy.tests.util`? You can use it as a context manager:
```
mykb = KnowledgeBase(nlp.vocab, entity_vector_length=1)
with make_tempdir() as d:
    mykb.to_disk(d / "kb")
    mykb.from_disk(d / "kb")
```

**Reply by graue70** [reply]
I thought about that but I don't think it helps. I need the temporary directory to exist not only during creation of something but also when I call the pipeline since the custom component uses files in the temp dir in its `__call__` method. So when the `@spacy.registry.misc` function returns an instance of `EntityIndex` the temporary directory needs to keep existing for the actual unit tests. Basically, it needs to exist for the duration of the test session.
I'm guessing that spacy doesn't know about the test session so I'm also guessing that it can't be achieved with spacy functionality like `make_tempdir()`. Am I wrong?

If you still think that `make_tempdir()` might help me please elaborate.

**Comment by graue70** [comment, marked as answer]
I found a solution:

```python
@pytest.fixture(scope="module")
def register_entity_index(tmp_path_factory):
    @spacy.registry.misc("test_entity_index")
    def entity_index():
        tmp_path = tmp_path_factory.mktemp("entity_index")
        db_path = tmp_path / "db.rocksdb"
        index = EntityIndex(db_path=db_path)
        index.build_from_csv(csv_path=Path("tests/entities.csv"))
        return index


@pytest.fixture(scope="module")
def register_entity_linker(register_entity_index):
    @Language.factory(
        "test_entity_linker",
        default_config={"entity_index": {"@misc": "test_entity_index"}},
    )
    def entity_linker(nlp, name, entity_index):
    return EntityLinker(entity_index=entity_index)


@pytest.fixture(scope="module")
def linker_pipeline(register_entity_linker):
    nlp = spacy.load("en_core_web_sm")
    nlp.add_pipe("test_entity_linker")
    return nlp
```

I basically realized that I could define a factory INSIDE a fixture which makes it possible to use another fixture inside the factory.
